php-checks/src/main/java/org/sonar/php/checks/UndocumentedApiCheck.java

```diff
@@ -0,0 +1,211 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.php.checks;
+
+import com.google.common.io.Files;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.php.api.CharsetAwareVisitor;
+import org.sonar.plugins.php.api.tree.declaration.ClassDeclarationTree;
+import org.sonar.plugins.php.api.tree.declaration.FunctionDeclarationTree;
+import org.sonar.plugins.php.api.tree.declaration.MethodDeclarationTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.List;
+
+/**
+ * Checks every method and class declaration for presence of a comment in the line before
+ * <p>
+ * Every declaration should have a short comment explaining what the {class|method|function} is doing
+ *
+ * @author Nils-Janis Mahlstädt <nils-janis.mahlstaedt@hmmh.de>
+ */
+@Rule(
+        key = UndocumentedApiCheck.KEY,
+        name = UndocumentedApiCheck.MESSAGE,
+        priority = Priority.MINOR,
+        tags = {Tags.DESIGN, Tags.CONVENTION})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.UNDERSTANDABILITY)
+@SqaleConstantRemediation("10min")
+public class UndocumentedApiCheck extends PHPVisitorCheck implements CharsetAwareVisitor {
+
+    public static final String KEY = "S1176";
+    public static final String MESSAGE = "method, function and class declarations should always be documented";
+
+    private static final String MESSAGE_PARTIAL = " declaration should always be documented";
+    private static final boolean defaultTrue = true;
+    private static final boolean defaultFalse = false;
+
+
+    private Charset charset;
+    private String currentFilename = "";
+    //lines in currentFile
+    private List<String> internal_lines;
+
+    // check if class has comment
+    @RuleProperty(
+            key = "checkClassForComment",
+            defaultValue = "true")
+    boolean checkClassComment = defaultTrue;
+
+    // check if function has comment
+    @RuleProperty(
+            key = "checkFunctionForComment",
+            defaultValue = "true")
+    boolean checkFunctionComment = defaultTrue;
+
+    // check if method has comment
+    @RuleProperty(
+            key = "checkMethodForComment",
+            defaultValue = "true")
+    boolean checkMethodComment = defaultTrue;
+
+    @Override
+    public void setCharset(Charset charset) {
+        this.charset = charset;
+    }
+
+
+    @Override
+    public void visitClassDeclaration(ClassDeclarationTree tree) {
```

Comment:
@nilsmahlstaedt You should simplify your implementation:

syntax tree already keeps information about comments: comments are saved for the token which follows this comments (see https://github.com/SonarSource/sonar-php/blob/master/php-frontend/src/main/java/org/sonar/plugins/php/api/tree/lexical/SyntaxToken.java#L36 trivia is the synonym for comment here)

So what you need - for each kind of declaration just find the first token of it and get the trivias/comments list for it. For each trivia you can get its value to check for `/**`